core/node/state_keeper/src/batch_executor/traits.rs

```diff
@@ -0,0 +1,238 @@
+//! Helper traits for the main executor.
+
+use std::{fmt, sync::Arc};
+
+use once_cell::sync::OnceCell;
+use zksync_multivm::{
+    interface::{
+        ExecutionResult, FinishedL1Batch, Halt, L1BatchEnv, L2BlockEnv, SystemEnv,
+        VmExecutionResultAndLogs, VmFactory, VmInterface, VmInterfaceHistoryEnabled,
+    },
+    tracers::CallTracer,
+    vm_latest::HistoryEnabled,
+    MultiVMTracer, VmInstance,
+};
+use zksync_state::{ReadStorage, StoragePtr, StorageView};
+use zksync_types::{vm_trace::Call, Transaction};
+use zksync_utils::bytecode::CompressedBytecodeInfo;
+
+use crate::{ExecutionMetricsForCriteria, TxExecutionResult};
+
+#[derive(Debug, Clone, Copy)]
+pub(super) enum TraceCalls {
+    Trace,
+    Skip,
+}
+
+#[derive(Debug)]
+pub(super) struct VmTransactionOutput {
+    tx_result: VmExecutionResultAndLogs,
+    compressed_bytecodes: Vec<CompressedBytecodeInfo>,
+    /// Empty if call tracing was not requested.
+    call_tracer_result: Vec<Call>,
+    gas_remaining: u32,
+}
+
+impl VmTransactionOutput {
+    fn from_err(err: VmTransactionError) -> Self {
+        let mut tx_result = *err.tx_result;
+        tx_result.result = ExecutionResult::Halt {
+            reason: Halt::FailedToPublishCompressedBytecodes,
+        };
+        Self {
+            tx_result,
+            compressed_bytecodes: vec![],
+            call_tracer_result: vec![],
+            gas_remaining: 0, // is not used; see below
+        }
+    }
+
+    pub(super) fn into_tx_result(self, tx: &Transaction) -> TxExecutionResult {
+        if let ExecutionResult::Halt { reason } = self.tx_result.result {
+            return match reason {
+                Halt::BootloaderOutOfGas => TxExecutionResult::BootloaderOutOfGasForTx,
+                _ => TxExecutionResult::RejectedByVm { reason },
+            };
+        }
+
+        let tx_metrics = ExecutionMetricsForCriteria::new(Some(tx), &self.tx_result);
+
+        TxExecutionResult::Success {
+            tx_result: Box::new(self.tx_result),
+            tx_metrics: Box::new(tx_metrics),
+            compressed_bytecodes: self.compressed_bytecodes,
+            call_tracer_result: self.call_tracer_result,
+            gas_remaining: self.gas_remaining,
+        }
+    }
+}
+
+#[derive(Debug)]
+pub(super) struct VmTransactionError {
+    tx_result: Box<VmExecutionResultAndLogs>,
+}
+
+/// Object-safe version of `VmInterface` used for batch execution.
+///
+/// # Invariants
+///
+/// - Transaction inspection methods must create a VM snapshot before the execution, and must not roll back these snapshots.
+/// - All rollbacks happen via `rollback_last_transaction`.
+pub(super) trait BatchVm {
```

Comment:
Not sure whether `BatchVm` and `BatchVmFactory` abstractions are warranted given that they are only used only for tests (for now?). IMO, `BatchVm` makes it easier to reason about batch executor – VM interactions and better structures code.

Reply:
The original intention was to make `BatchExecutor` trait to be responsible for that kind of stuff (back then the hypothesis was that the new VM will not be lifetime-bound, so we won't have to spawn a dedicated thread for each batch). I'm not against having `BatchVm`/`BatchVmFactory` per se, but then I think we should remove `BatchExecutor` trait (looks like we only have a single alternative implementation for tests, so hopefully it shouldn't be too hard).